# **Partner Business Modeling**

## Description

This data project has been used as a take-home assignment in the recruitment process for the data science positions at Uber.

## Assignment

***Scenario 1:*** *It is going to be a huge Saturday and there will need to be many more cars on the road than last week. In order to get drivers to go online, we're assessing the following two bonus options in terms of cost:*

- *Option 1: $50 for each driver that is online at least 8 hours, accepts 90% of requests, completes 10 trips, and has a rating of 4.7 or better during the time frame;*
- *Option 2: $4/trip for all drivers who complete 12 trips, and have a 4.7 or better rating.*

Using the dataset provided and given Scenario 1, provide answers to the questions below:

1. How much would the total bonus payout be with Option 1?
2. How much would the total bonus payout be with Option 2?
3. How many drivers would qualify for a bonus under Option 1 but not under Option 2?
4. What percentages of drivers online completed less than 10 trips, had an acceptance rate of less than 90%, and had a rating of 4.7 or higher?

***Scenario 2:*** *A taxi driver currently generates $200 per day in fares (before expenses), works six days a week, takes three weeks off, and has the following expenses:*

- *Gas - $200 per week*
- *Insurance - $400 per month*
- *Vehicle rent (by the week) - $500*

*The driver doesn't pay gas and rent expenses on off weeks.*

*Now, let's assume that the same driver would buy a Town Car and partner with Uber. If he does, his gas expenses would go up by 5%, his insurance expense would decrease by 20%, and he would no longer be renting a vehicle. However, he would need to buy a car. The driver would still take three weeks off per year.*

Given Scenario 2, provide answers to the questions below:

5. How much money (after expenses) does the taxi driver make per year without partnering with Uber?
6. You are convincing the same driver above to buy a Town Car and partner with Uber. Assuming the new car is 40,000 USD, how much would the driver's gross fares need to increase per week to fully pay for the car in year 1 and maintain the same yearly profit margin as before?


## Practicalities

Please work on the questions in the displayed order. Make sure that the solution reflects your entire thought process - it is more important how the code is structured rather than the final answers. You are expected to spend no more than 2 hours solving this project.


## Data Exploration

In [3]:
import pandas as pd
import numpy as np

df = pd.read_csv('datasets/dataset_2.csv')
df.head(10)

,Name,Trips Completed,Accept Rate,Supply Hours,Rating
0,Abdul,1,100%,3,4.8
1,Abraham,12,83%,5,4.7
2,Adelina,1,100%,2,4.7
3,Akilah,1,100%,2,4.9
4,Alec,21,76%,11,5.0
5,Alesha,7,100%,4,4.8
6,Alvaro,17,88%,11,4.6
7,Andra,16,94%,11,4.6
8,Augusta,19,84%,11,4.7
9,Aurora,10,90%,4,4.6


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119 entries, 0 to 118
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             119 non-null    object 
 1   Trips Completed  119 non-null    int64  
 2   Accept Rate      119 non-null    object 
 3   Supply Hours     119 non-null    int64  
 4   Rating           119 non-null    float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.8+ KB


In [5]:
df.shape

(119, 5)

## Question 1

It is going to be a huge Saturday and there will need to be many more cars on the road than last week. In order to get drivers to go online, we're assessing the following two bonus options in terms of cost:

- Option 1: $50 for each driver that is online at least 8 hours, accepts 90\% of requests, completes 10 trips, and has a rating of 4.7 or better during the time frame.
- Option 2: $4 trip for all drivers who complete 12 trips, and have a 4.7 or better rating.

**How much would the total bonus payout be with Option 1?**

In [6]:
df['Accept Rate'] = df['Accept Rate'].apply(lambda x: float(x[:-1]))
df

,Name,Trips Completed,Accept Rate,Supply Hours,Rating
0,Abdul,1,100.0,3,4.8
1,Abraham,12,83.0,5,4.7
2,Adelina,1,100.0,2,4.7
3,Akilah,1,100.0,2,4.9
4,Alec,21,76.0,11,5.0
...,...,...,...,...,...
114,Virgen,4,100.0,6,4.5
115,Yang,7,71.0,2,4.5
116,Yessenia,8,88.0,5,5.0
117,Yukiko,9,78.0,6,4.5


In [7]:
first_option_df = df[(df['Supply Hours'] >= 8) & (df['Trips Completed'] >= 10) & 
                     (df['Accept Rate'] >= 90) & (df['Rating'] >= 4.7)]
first_option_df

,Name,Trips Completed,Accept Rate,Supply Hours,Rating
11,Byron,15,100.0,10,4.9
12,Carlota,14,100.0,8,5.0
19,Dannette,14,100.0,9,4.9
23,Demetrius,14,100.0,9,5.0
26,Dimple,15,100.0,10,4.9
32,Emil,12,100.0,9,5.0
37,Garth,15,100.0,10,5.0
40,Hanh,14,94.0,9,4.9
53,Keshia,20,100.0,11,4.8
57,Latonia,13,100.0,9,5.0


In [8]:
first_option_total_payout = 50 * len(first_option_df)
print('$'+str(first_option_total_payout))

$1050


**$1050** is the total payout for Option 1

## Question 2

**How much would the total bonus payout be with option 2?**

In [9]:
second_option_df = df[(df['Trips Completed'] >= 12) & (df['Rating'] >= 4.7)]
second_option_df

,Name,Trips Completed,Accept Rate,Supply Hours,Rating
1,Abraham,12,83.0,5,4.7
4,Alec,21,76.0,11,5.0
8,Augusta,19,84.0,11,4.7
10,Buffy,13,54.0,6,5.0
11,Byron,15,100.0,10,4.9
12,Carlota,14,100.0,8,5.0
15,Chu,14,71.0,7,4.8
19,Dannette,14,100.0,9,4.9
21,Deane,22,77.0,9,4.7
23,Demetrius,14,100.0,9,5.0


In [1]:
second_option_total_payout = 4 * second_option_df['Trips Completed'].sum()
print('$'+str(second_option_total_payout))

NameError: name 'second_option_df' is not defined

**$2976** is the total payout bonus for Payout 2

## Question 3

**How many drivers would qualify under Option 1 but not under Option 2?**

In [13]:
df_all = first_option_df.merge(second_option_df, on=first_option_df.columns.to_list(), 
                   how='left', indicator=True)
df_all

,Name,Trips Completed,Accept Rate,Supply Hours,Rating,_merge
0,Byron,15,100.0,10,4.9,both
1,Carlota,14,100.0,8,5.0,both
2,Dannette,14,100.0,9,4.9,both
3,Demetrius,14,100.0,9,5.0,both
4,Dimple,15,100.0,10,4.9,both
5,Emil,12,100.0,9,5.0,both
6,Garth,15,100.0,10,5.0,both
7,Hanh,14,94.0,9,4.9,both
8,Keshia,20,100.0,11,4.8,both
9,Latonia,13,100.0,9,5.0,both


In [14]:
df_all[df_all["_merge"] == 'left_only']

,Name,Trips Completed,Accept Rate,Supply Hours,Rating,_merge
16,Oren,11,91.0,9,4.8,left_only
17,Phyllis,10,90.0,8,4.8,left_only


## Question 4

**What percentages of drivers online completed less than 10 trips, had an acceptance rate of less than 90%, and had a rating of 4.7 or higher?**

In [15]:
less_trips_df = df[(df['Trips Completed'] < 10) & (df['Accept Rate'] < 90) & (df['Rating'] >= 4.7)]
less_trips_df

,Name,Trips Completed,Accept Rate,Supply Hours,Rating
17,Cris,7,71.0,5,5.0
18,Dalila,7,57.0,2,4.8
20,Dannielle,3,67.0,5,5.0
22,Delfina,4,50.0,3,4.7
27,Domenica,9,89.0,5,4.9
36,Floyd,3,67.0,1,4.8
41,Hee,9,89.0,7,4.7
45,Ingrid,7,43.0,4,4.8
66,Lilla,9,89.0,8,4.7
67,Loree,9,89.0,8,4.7


In [16]:
print(str(len(less_trips_df) / len(df) * 100)+"%")

10.92436974789916%


## Question 5

A taxi driver currently generates $200 per day in fares (before expenses), works six days a week, takes three weeks off, and has the following expenses:

- Gas: 200 USD per week
- Insurance: 400 USD per month
- Vehicle rent (by the week): 500 USD

The driver doesn't pay gas and rent expenses on off weeks.

**How much money (after expenses) does the driver make per year?**

In [17]:
total_weeks_per_year = 52
weeks_off = 3
fare_per_day = 200 
workday_per_week = 6
total_months_per_year = 12

# expenses
gas_per_week = 200
insurance_per_month = 400
vehicle_rent_by_week = 500

total_expenses = (gas_per_week + vehicle_rent_by_week) * (total_weeks_per_year - weeks_off) + insurance_per_month * total_months_per_year
print("$"+str(total_expenses))

$39100


In [18]:
total_revenue = (total_weeks_per_year - weeks_off) * workday_per_week * fare_per_day
print("$"+str(total_revenue))

$58800


In [19]:
profit_margin = (total_revenue - total_expenses) / total_revenue
print(str(round(profit_margin * 100, 2)) + "%")

33.5%


## Question 6

*You are convincing the same driver above to buy a Town Car and partner with Uber.*

*If he does, his gas expenses would go up by 5%, his insurance expense would decrease by 20%, and he would no longer be renting a vehicle. However, he would need to buy a car. The driver would still take three weeks off per year.*

**Assuming the new car is 40,000 USD, how much would the driver's gross fares need to increase per week to fully pay for the car in year 1 and maintain the same yearly profit margin as before?**

In [20]:
# calculate new expenses
gas_per_week = gas_per_week * 1.05
insurance_per_month = insurance_per_month * 0.8
new_car = 40000

new_total_expenses = new_car + gas_per_week * (total_weeks_per_year - weeks_off) + insurance_per_month * total_months_per_year
print("$"+str(new_total_expenses))

$54130.0


In [21]:
# profit margin is the total revenue from previous question minus new expenses
new_profit_margin = (total_revenue - new_total_expenses) / total_revenue
print(str(round(new_profit_margin * 100, 2)) + "%")

7.94%


In [22]:
# calculate the weekly fare increase
fare_increase = total_revenue * (profit_margin - new_profit_margin) / (total_weeks_per_year - weeks_off)
print("$"+str(round(fare_increase, 2)))

$306.73
